src/flatten.jl

```diff
@@ -17,13 +17,19 @@ true
 """
 function flatten end
 
-function flatten(x::Real)
+function flatten(x::Integer)
+    v = Float64[]
+    unflatten_Integer(v::Vector{Float64}) = x
+    return v, unflatten_Integer
+end
+
+function flatten(x::AbstractFloat)
     v = [x]
     unflatten_to_Real(v::Vector{<:Real}) = only(v)
     return v, unflatten_to_Real
 end
 
-flatten(x::Vector{<:Real}) = (x, identity)
+flatten(x::Vector{<:AbstractFloat}) = (x, identity)
```

Comment:
just wonder why `flatten(x::Vector{<:AbstractFloat})` is required? It will go through `flatten(x::AbstractVector)` and then `flatten(x::AbstractFloat)`, right? In other words, if `flatten(x::Vector{<:AbstractFloat})` is required, by not `flatten(x::Vector{<:Int})`?

Reply:
Yup. This is just an optimisation

Reply:
sorry, what do you mean?

Reply:
You're correct that `flatten(x::Vector{<:AbstractFloat})` isn't required for correctness -- as you say, if it weren't there then everything would just recurse and produce the same result. In that sense, it's just an optimisation to make this particular method of `flatten` be fast.

You could make the same argument for the implementation of `flatten` for vectors of integers, but it doesn't seem like a priority in the same way that `Vector{<:AbstractFloat}` is.